pkg/reconciler/managed/reconciler.go

```diff
@@ -1248,3 +1291,92 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
 	managed.SetConditions(xpv1.ReconcileSuccess())
 	return reconcile.Result{RequeueAfter: reconcileAfter}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 }
+
+// recordChangeLog constructs a change log entry from the provided information
+// and sends it to the change log service, if the feature is enabled. If errors
+// are encountered during this process, then logs/events will be recorded for
+// each, but the process will continue. We are choosing to send all the
+// information we can to the change logs as opposed to sending nothing.
+func (r *Reconciler) recordChangeLog(ctx context.Context, managed resource.Managed, snapshot *structpb.Struct, log logging.Logger,
+	record event.Recorder, opType changelogs.OperationType, changeErr error, ad *AdditionalDetails,
+) {
```

Comment:
This function signature is pretty long for my comfort.

https://github.com/crossplane/crossplane/tree/master/contributing#dont-wrap-function-signatures

I wonder, does this need to be a method on `Reconciler`? It seems to have most of its dependencies passed in as arguments. I suggest making it a self-contained injectable type that wraps the 'raw' `changelogs.ChangeLogServiceClient` gRPC generates.

Something like this:

```go
type ChangeLogger struct {
  client changelings.ChangeLogServiceClient
}

func (c *ChangeLogger) RecordChangeLog(ctx context.Context, mg resource.Managed, cl *changelogs.ChangeLog) error {
  // Logic goes here
}
```

I would then:

* Not pass a logger and event recorder to this type. Instead, return descriptive errors up to the `Reconciler` and let that log the errors and/or create events.
* Define a `ChangeLog` message at the Protobuf level, so that snapshot, opType, and ad all become part of one message you can pass into `RecordChangeLog`.

Reply:
I've made a bunch of data format changes to the protobuf types as suggested, and I'm starting to look into this now.  I also already thought the `recordChangeLog()` method signature was overly long, so I'm not against exploring this suggestion too.

However, one thing I liked about my first implementation was that everything was encapsulated in the `recordChangeLog` method, including change log entry preparation, error handling, and event recording. This allowed the `Reconcile()` function, which is pretty complex and long, to just get a single line addition where needed to call `recordChangeLog`. With the approach you suggest, I believe we'll have additional lines within `Reconcile()` to prep the change log entry with data and to handle errors / record events in all the places we call it (create/update/delete - success/failure).

That's the downside I see to this approach, but it may end up being palatable.  I will play with it and see! 🤓

Reply:
Took a quick stab at this with https://github.com/crossplane/crossplane-runtime/pull/754/commits/ae4de71d3d844f1697530e4af8a40810ce09ea41. The change logs functionality has been broken out from the managed reconciler into a `changeLogger` type with a few key methods:

* `takeSnapshot()` to capture the state of the managed resource early during reconciliation before any operations are performed
* `newChangeLogEntry()` to create and prep a change log entry with given values/details
* `recordChangeLog()` to actually send the change log entry to the underlying change log service client

Note that the error handling and event recording is being done inside this `changeLogger`, so the main `Reconcile()` doesn't get many additional lines added. All that `Reconcile()` needs to have added is basically `newChangeLogEntry(a, b, c)` followed by `recordChangeLog(entry)`.

I think I like keeping changes to `Reconcile()` very minimal, but still playing with this idea a bit 😇

Reply:
Addressed in a more cleanly designed way as discussed in https://github.com/crossplane/crossplane-runtime/pull/754/commits/c23f09b8314e5a7d11e0438243572a0b38cfaff1